### ISSUE TITLE
CRD error when updating from 0.36 to 0.37

### ISSUE THREAD

**Issue by brew** [fluxcd/flux2#3362]
### Describe the bug

After successfully updating from 0.31 to 0.36, I ran into a problem updating to 0.37. 

If I use `flux bootstrap ...` to update flux from 0.36 to 0.37 I receive the following error:

```
➜ flux bootstrap github \
    --components=source-controller,kustomize-controller,helm-controller,notification-controller \
    --components-extra=image-reflector-controller,image-automation-controller \
    --path=k8s/cluster \
    --owner=<owner> \
    --repository=<repo> \
    --branch=master \
    --read-write-key
► connecting to github.com
► cloning branch "master" from Git repository "https://github.com/<owner>/<repo>.git"
✔ cloned repository
► generating component manifests
✔ generated component manifests
✔ committed sync manifests to "master" ("<commit hash>")
► pushing component manifests to "https://github.com/<owner>/<repo>.git"
✔ reconciled components
► determining if source secret "flux-system/flux-system" exists
✔ source secret up to date
► generating sync manifests
✔ generated sync manifests
✔ sync manifests are up to date
► applying sync manifests
✗ CustomResourceDefinition/imagepolicies.image.toolkit.fluxcd.io dry-run failed, reason: Invalid, error: CustomResourceDefinition.apiextensions.k8s.io "imagepolicies.image.toolkit.fluxcd.io" is invalid: status.storedVersions[0]: Invalid value: "v1alpha2": must appear in spec.versions
```

The imagepolicies CRD returned from the cluster:

```
➜ kubectl get crd imagepolicies.image.toolkit.fluxcd.io -o yaml

apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  annotations:
    controller-gen.kubebuilder.io/version: v0.8.0
  creationTimestamp: "2021-06-07T09:10:21Z"
  generation: 6
  labels:
    app.kubernetes.io/component: image-reflector-controller
    app.kubernetes.io/instance: flux-system
    app.kubernetes.io/part-of: flux
    app.kubernetes.io/version: v0.36.0
    kustomize.toolkit.fluxcd.io/name: flux-system
    kustomize.toolkit.fluxcd.io/namespace: flux-system
  name: imagepolicies.image.toolkit.fluxcd.io
  resourceVersion: "423639169"
  uid: fd80ce48-b0a6-4c9d-a0c7-cb6c35304f01
spec:
  conversion:
    strategy: None
  group: image.toolkit.fluxcd.io
  names:
    kind: ImagePolicy
    listKind: ImagePolicyList
    plural: imagepolicies
    singular: imagepolicy
  scope: Namespaced
  versions:
  - additionalPrinterColumns:
    - jsonPath: .status.latestImage
      name: LatestImage
      type: string
    name: v1alpha1
    schema:
      ...
    served: true
    storage: false
    subresources:
      status: {}
  - additionalPrinterColumns:
    - jsonPath: .status.latestImage
      name: LatestImage
      type: string
    name: v1alpha2
    schema:
      ...
    served: true
    storage: false
    subresources:
      status: {}
  - additionalPrinterColumns:
    - jsonPath: .status.latestImage
      name: LatestImage
      type: string
    name: v1beta1
    schema:
      ...
    served: true
    storage: true
    subresources:
      status: {}
status:
  acceptedNames:
    kind: ImagePolicy
    listKind: ImagePolicyList
    plural: imagepolicies
    singular: imagepolicy
  conditions:
  - lastTransitionTime: "2021-06-07T09:10:21Z"
    message: no conflicts found
    reason: NoConflicts
    status: "True"
    type: NamesAccepted
  - lastTransitionTime: "2021-06-07T09:10:21Z"
    message: the initial names have been accepted
    reason: InitialNamesAccepted
    status: "True"
    type: Established
  storedVersions:
  - v1alpha2
  - v1beta1

```

### Steps to reproduce

1. Install flux with `brew install fluxcd/tap/flux`
2. Run bootstrap:

```
flux bootstrap github \
    --components=source-controller,kustomize-controller,helm-controller,notification-controller \
    --components-extra=image-reflector-controller,image-automation-controller \
    --path=k8s/cluster \
    --owner=<owner> \
    --repository=<repo> \
    --branch=master \
    --read-write-key
```

### Expected behavior

Expect bootstrap to finish successfully and for the CRD to be updated to 0.37.

### OS / Distro

macOS 12.6.1

### Flux version

flux: v0.37.0

### Flux check

► checking prerequisites
✔ Kubernetes 1.21.14-eks-fb459a0 >=1.20.6-0
► checking controllers
✔ helm-controller: deployment ready
► ghcr.io/fluxcd/helm-controller:v0.26.0
✔ image-automation-controller: deployment ready
► ghcr.io/fluxcd/image-automation-controller:v0.26.1
✔ image-reflector-controller: deployment ready
► ghcr.io/fluxcd/image-reflector-controller:v0.22.1
✔ kustomize-controller: deployment ready
► ghcr.io/fluxcd/kustomize-controller:v0.30.0
✔ notification-controller: deployment ready
► ghcr.io/fluxcd/notification-controller:v0.28.0
✔ source-controller: deployment ready
► ghcr.io/fluxcd/source-controller:v0.31.0
► checking crds
✔ alerts.notification.toolkit.fluxcd.io/v1beta1
✔ buckets.source.toolkit.fluxcd.io/v1beta1
✔ gitrepositories.source.toolkit.fluxcd.io/v1beta1
✔ helmcharts.source.toolkit.fluxcd.io/v1beta1
✔ helmreleases.helm.toolkit.fluxcd.io/v2beta1
✔ helmrepositories.source.toolkit.fluxcd.io/v1beta1
✔ imagepolicies.image.toolkit.fluxcd.io/v1alpha2
✔ imagerepositories.image.toolkit.fluxcd.io/v1alpha2
✔ imageupdateautomations.image.toolkit.fluxcd.io/v1alpha2
✔ kustomizations.kustomize.toolkit.fluxcd.io/v1beta1
✔ ocirepositories.source.toolkit.fluxcd.io/v1beta2
✔ providers.notification.toolkit.fluxcd.io/v1beta1
✔ receivers.notification.toolkit.fluxcd.io/v1beta1
✔ all checks passed

### Git provider

GitHub

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by stefanprodan** [comment]
The v1alpha2 APIs have been removed as they were deprecated one year ago. Update your definitions to their beta version in Git

**Comment by pjbgf** [comment]
@brew thank you for reporting this.

As part of `v0.37.0` we have dropped support for alpha API versions for both [Image Automation Controller](https://github.com/fluxcd/image-automation-controller/pull/456) and [Image Reflector Controller](https://github.com/fluxcd/image-reflector-controller/pull/323). Those APIs were deprecated for over a year now.

Please upgrade your YAML files to refer to the latest API versions instead, which are `v1beta1` for both controllers.
Here's some information around the moves and changes across the APIs:
https://fluxcd.io/flux/components/image/imageupdateautomations/#moves-and-changes

**Comment by brew** [comment]
Sorry, I should have mentioned, I have updated the ImagePolicy, ImageUpdateAutomation, and ImageRepository definitions to `image.toolkit.fluxcd.io/v1beta1`.

**Comment by brew** [comment]
Is this error because the objects stored in etcd are using the old api? 

Although my manifests in git use `v1beta1`, I can see (using the AWS EKS web console, which I assume is a view of the stored object in etcd?) that the Raw view of my ImagePolicy objects is using `v1alpha2`. Is it not enough to update my manifests in git to the updated api version? How can I update what's stored in etcd?

**Comment by brew** [comment]
I uninstalled the CRDs by running `flux uninstall` and re-ran bootstrap to reinstall them to refresh etcd. So glad `flux uninstall` is a safe operation 😄.

**Comment by pjbgf** [comment]
Closing the issue and considering `flux uninstall` before a new bootstrap as the workaround to be used in such cases.

**Comment by SerialVelocity** [comment]
So, this is because if you had the old version in the past, it sometimes doesn't get removed from the CRD. What you can do though is run:
```
kubectl edit crd imagepolicies.image.toolkit.fluxcd.io --subresource status
kubectl edit crd imagerepositories.image.toolkit.fluxcd.io --subresource status
```
and manually remove it from the `storedVersions` block at the bottom. You must use `--subresource` or it cannot be modified.

**Comment by SerialVelocity** [comment]
The `--subresource- flag has been merged since kubectl 1.24.0. It's probably best to avoid the alpha binary just in case.

Issue about it here:
https://github.com/kubernetes/enhancements/issues/2590

**Comment by uluzox** [comment]
I also ran into this issue.
We use a helm chart for all image.toolkit.fluxcd.io resources. This I upgraded to use the newer api version and spec (if necessary) before the fluxCD upgrade. 
However, somehow whenever the HelmRelease creates a i.e. ImagePolicy of v1beta1, fluxCd would also create one of version v1alpha1. Both resources where identical even in uuid but for the apiVersion. But since v1alpha1 has a different spec, it got created in a broken state and was undeletable.

Flux self-update was unable to apply the new version with
> CustomResourceDefinition/imagepolicies.image.toolkit.fluxcd.io dry-run failed, reason: Invalid, error: CustomResourceDefinition.apiextensions.k8s.io "imagepolicies.image.toolkit.fluxcd.io" is invalid: [status.storedVersions[0]: Invalid value: "v1alpha1": must appear in spec.versions, status.storedVersions[1]: Invalid value: "v1alpha2": must appear in spec.versions]


**The fix**

1. Suspend cluster kustomization
2. Delete all HelmReleases that would otherwise recreate flux image resources
3. Now HR are stuck in "to-be-deleted" because flux image* resources cannot be deleted
4. Delete all alpha and beta image* resources
5. Now they are stuck in "to-be-deleted"
6. Patch all alpha resources to conform to alpha CRD
7. Now the deletion process kicks in and removes them and also the beta ones and ultimately also the HR
8. Do [this](https://github.com/fluxcd/flux2/issues/3362#issuecomment-1353240741) for all image*.image.toolkit.fluxcd.io resources
9. Wait for flux-self-update to reconcile the next time (or manually trigger it)
10. Now flux upgrade was successfulResume cluster kustomization -> recreates HR -> recreates flux image resources
